README.md
# NAME

Set::Similarity - similarity measures for sets

<div>
    <a href="https://travis-ci.org/wollmers/Set-Similarity"><img src="https://travis-ci.org/wollmers/Set-Similarity.png" alt="Set-Similarity"></a>
    <a href='https://coveralls.io/r/wollmers/Set-Similarity?branch=master'><img src='https://coveralls.io/repos/wollmers/Set-Similarity/badge.png?branch=master' alt='Coverage Status' /></a>
    <a href='http://cpants.cpanauthors.org/dist/Set-Similarity'><img src='http://cpants.cpanauthors.org/dist/Set-Similarity.png' alt='Kwalitee Score' /></a>
    <a href="http://badge.fury.io/pl/Set-Similarity"><img src="https://badge.fury.io/pl/Set-Similarity.svg" alt="CPAN version" height="18"></a>
</div>

# SYNOPSIS

    use Set::Similarity::Dice;

    # object method
    my $dice = Set::Similarity::Dice->new;
    my $similarity = $dice->similarity('Photographer','Fotograf');

    # class method
    my $dice = 'Set::Similarity::Dice';
    my $similarity = $dice->similarity('Photographer','Fotograf');

    # from 2-grams
    my $width = 2;
    my $similarity = $dice->similarity('Photographer','Fotograf',$width);

    # from arrayref of tokens
    my $similarity = $dice->similarity(['a','b'],['b']);

    # from hashref of features
    my $bird = {
      wings    => true,
      eyes     => true,
      feathers => true,
      hairs    => false,
      legs     => true,
      arms     => false,
    };
    my $mammal = {
      wings    => false,
      eyes     => true,
      feathers => false,
      hairs    => true,
      legs     => true,
      arms     => true,
    };
    my $similarity = $dice->similarity($bird,$mammal);

    # from arrayref sets
    my $bird = [qw(
      wings
      eyes
      feathers
      legs
    )];
    my $mammal = [qw(
      eyes
      hairs
      legs
      arms
    )];
    my $similarity = $dice->from_sets($bird,$mammal);

# DESCRIPTION

This is the base class including mainly helper and convenience methods.

## Overlap coefficient

( A intersect B ) / min(A,B)

## Jaccard Index

The Jaccard coefficient measures similarity between sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets

( A intersect B ) / (A union B)

The Tanimoto coefficient is the ratio of the number of features common to both sets to the total number of features, i.e.

( A intersect B ) / ( A + B - ( A intersect B ) ) # the same as Jaccard

The range is 0 to 1 inclusive.

## Dice coefficient

The Dice coefficient is the number of features in common to both sets relative to the average size of the total number of features present, i.e.

( A intersect B ) / 0.5 ( A + B ) # the same as sorensen

The weighting factor comes from the 0.5 in the denominator. The range is 0 to 1.

# METHODS

All methods can be used as class or object methods.

## new

    $object = Set::Similarity->new();

## similarity

    my $similarity = $object->similarity($any1,$any1,$width);

`$any` can be an arrayref, a hashref or a string. Strings are tokenized into n-grams of width `$width`.

`$width` must be integer, or defaults to 1.

## from\_tokens

    my $similarity = $object->from_tokens(['a','b'],['b']);

## from\_sets

    my $similarity = $object->from_sets(['a'],['b']);

Croaks if called directly. This method should be implemented in a child module.

## intersection

    my $intersection_size = $object->intersection(['a'],['b']);

## uniq

    my @uniq = $object->uniq(['a','b']);

Transforms an arrayref of strings into an array of unique elements.

## combined\_length

    my $set_size_sum = $object->combined_length(['a'],['b']);

## min

    my $min_set_size = $object->min(['a'],['b']);

## ngrams

    my @monograms = $object->ngrams('abc');
    my @bigrams = $object->ngrams('abc',2);

## \_any

    my $arrayref = $object->_any($any,$width);

# SEE ALSO

[Set::Similarity::Cosine](https://metacpan.org/pod/Set%3A%3ASimilarity%3A%3ACosine)

[Set::Similarity::Dice](https://metacpan.org/pod/Set%3A%3ASimilarity%3A%3ADice)

[Set::Similarity::Jaccard](https://metacpan.org/pod/Set%3A%3ASimilarity%3A%3AJaccard)

[Set::Similarity::Overlap](https://metacpan.org/pod/Set%3A%3ASimilarity%3A%3AOverlap)

[Bag::Similarity](https://metacpan.org/pod/Bag%3A%3ASimilarity) doing the same for bags or multisets.

[Text::Levenshtein](https://metacpan.org/pod/Text%3A%3ALevenshtein) for distance measures of strings, and a very overview of similar modules,

[http://en.wikipedia.org/wiki/String\_metric](http://en.wikipedia.org/wiki/String_metric) for an overview of similarity measures.

[Cluster::Similarity](https://metacpan.org/pod/Cluster%3A%3ASimilarity) for clusters.

# SOURCE REPOSITORY

[http://github.com/wollmers/Set-Similarity](http://github.com/wollmers/Set-Similarity)

# AUTHOR

Helmut Wollmersdorfer, <helmut@wollmersdorfer.at>

<div>
    <a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>
</div>

# COPYRIGHT AND LICENSE

Copyright (C) 2013-2020 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Set::Similarity

{{$NEXT}}

0.027  2020-08-06 18:38:03 CEST
       - improve Kwalitee by provides

0.026  2016-11-24 12:04:49 CET
       - travis for Perl 5.24

0.025  2016-11-24 11:33:22 CET
       - requires 'parent'

0.024  2015-04-27 15:44:19 CEST
       - improved travis configuration and badges

0.023  2015-03-31 12:15:04 CEST
       - added badges

0.022  2014-10-09 14:38:46 CEST
       - improved test coverage

0.021  2014-10-09 13:36:47 CEST
       - improved test and pod coverage
       - improved documentation

0.020  2014-08-20 20:01:02 CEST
       - remove n-gram padding, return the unchanged word if shorter

0.019  2014-08-20 09:04:46 CEST
       - pad if length word < n-gram width

0.018  2014-08-19 23:45:03 CEST
       - remove useless code, force 1-gram if width not integer

0.017  2014-08-16 23:43:23 CEST
       - badges in pod

0.016  2014-08-16 13:24:55 CEST
       - test croak

0.015  2014-08-16 07:47:26 CEST
       - add travis

0.014  2014-08-08 22:59:01 CEST
       - remove default method and croak

0.013  2014-08-04 14:26:05 CEST
       - repair versioning

0.012  2014-08-03 22:03:41 CEST
       - improved documentation

0.011  2014-08-02 11:17:41 CEST
       - change minimal Perl version to 5.6

0.010  2014-08-01 21:53:39 CEST
       - improved documentation

0.009  2014-07-27 08:16:29 CEST
       - improved documentation

0.008  2014-07-26 19:15:31 CEST
       - improved documentation

0.007  2014-07-26 18:14:04 CEST
       - use arrayref

0.006  2014-05-17 20:43:12 CEST
       - release a trial release

0.005  2014-05-17 15:32:39 CEST
       - try trial release

0.004  2014-05-17 15:05:39 CEST
       - improve docs

0.003  2014-04-30 21:59:32 CEST
       - remove mymeta

0.002  2014-04-22 22:44:52 CEST
       - Added Set::Similarity::Cosine

0.001  2014-03-18
       - Initial release.


lib/Set/Similarity/Dice.pm
package Set::Similarity::Dice;

use strict;
use warnings;

use parent 'Set::Similarity';

our $VERSION = '0.025';

sub from_sets {
  my ($self, $set1, $set2) = @_;

  # $dice = ($intersection * 2 / $combined_length);
  return (
    $self->intersection($set1,$set2) * 2 / $self->combined_length($set1,$set2)
  );
}

1;

__END__

=head1 NAME

Set::Similarity::Dice - Dice coefficent for sets

=head1 SYNOPSIS

 use Set::Similarity::Dice;
 
 my $dice = Set::Similarity::Dice->new;
 my $similarity = $dice->similarity('Photographer','Fotograf');
 
 
=head1 DESCRIPTION

=head2 Dice coefficient

The Dice coefficient is the number of elements in common to both sets relative to the 
average size of the total number of elements present, i.e.

( A intersect B ) / 0.5 ( A + B ) # the same as sorensen

The weighting factor comes from the 0.5 in the denominator. The range is 0 to 1.

=head1 METHODS

L<Set::Similarity::Dice> inherits all methods from L<Set::Similarity> and implements the
following new ones.

=head2 from_sets

  my $similarity = $object->from_sets(['a'],['b']);
 
This method expects two arrayrefs of strings as parameters. The parameters are not checked, thus can lead to funny results or uncatched divisions by zero.
 
If you want to use this method directly, you should take care that the elements are unique. Also you should catch the situation where one of the arrayrefs is empty (similarity is 0), or both are empty (similarity is 1).

=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Set-Similarity>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2014 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut




lib/Set/Similarity/Cosine.pm
package Set::Similarity::Cosine;

use strict;
use warnings;

use parent 'Set::Similarity';

our $VERSION = '0.025';

sub from_sets {
  my ($self, $set1, $set2) = @_;

  # it is so simple because the vectors contain only 0 and 1
  return (
    $self->intersection($set1,$set2) / (sqrt(scalar @$set1) * sqrt(scalar @$set2))
  );
}

1;

__END__

=head1 NAME

Set::Similarity::Cosine - Cosine similarity for sets

=head1 SYNOPSIS

 use Set::Similarity::Cosine;
 
 my $cosine = Set::Similarity::Cosine->new;
 my $similarity = $cosine->similarity('Photographer','Fotograf');
 

=head1 DESCRIPTION

=head2 Cosine similarity

A intersection B / (sqrt(A) * sqrt(B))


=head1 METHODS

L<Set::Similarity::Cosine> inherits all methods from L<Set::Similarity> and implements the
following new ones.

=head2 from_sets

  my $similarity = $object->from_sets(['a'],['b']);
 
This method expects two arrayrefs of strings as parameters. The parameters are not checked, thus can lead to funny results or uncatched divisions by zero.
 
If you want to use this method directly, you should take care that the elements are unique. Also you should catch the situation where one of the arrayrefs is empty (similarity is 0), or both are empty (similarity is 1).

=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Set-Similarity>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2014 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Set/Similarity/Jaccard.pm
package Set::Similarity::Jaccard;

use strict;
use warnings;

use parent 'Set::Similarity';

our $VERSION = '0.025';

sub from_sets {
  my ($self, $set1, $set2) = @_;

  my $intersection = $self->intersection($set1,$set2);
  my $union = $self->combined_length($set1,$set2) - $intersection;
  # ( A intersect B ) / (A union B)
  return ($intersection / $union);
}

1;

__END__

=head1 NAME

Set::Similarity::Jaccard - Jaccard coefficent for sets

=head1 SYNOPSIS

 use Set::Similarity::Jaccard;
 
 my $jaccard = Set::Similarity::Jaccard->new;
 my $similarity = $jaccard->similarity('Photographer','Fotograf');
 

=head1 DESCRIPTION

=head2 Jaccard Index

The Jaccard coefficient measures similarity between sample sets, and is defined as the 
size of the intersection divided by the size of the union of the sample sets

( A intersect B ) / (A union B)

The Tanimoto coefficient is the ratio of the number of elements common to both sets to 
the total number of elements, i.e.

( A intersect B ) / ( A + B - ( A intersect B ) ) # the same as Jaccard 

The range is 0 to 1 inclusive.

=head1 METHODS

L<Set::Similarity::Jaccard> inherits all methods from L<Set::Similarity> and implements the
following new ones.

=head2 from_sets

  my $similarity = $object->from_sets(['a'],['b']);

This method expects two arrayrefs of strings as parameters. The parameters are not checked, thus can lead to funny results or uncatched divisions by zero.
 
If you want to use this method directly, you should take care that the elements are unique. Also you should catch the situation where one of the arrayrefs is empty (similarity is 0), or both are empty (similarity is 1).
 
=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Set-Similarity>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2014 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/Set/Similarity/Overlap.pm
package Set::Similarity::Overlap;

use strict;
use warnings;

use parent 'Set::Similarity';

our $VERSION = '0.025';

sub from_sets {
  my ($self, $set1, $set2) = @_;

  # ( A intersect B ) / min(A,B)  
  return (
    $self->intersection($set1,$set2) / $self->min($set1,$set2)
  );
}

1;

__END__

=head1 NAME

Set::Similarity::Overlap - Overlap coefficent for sets

=head1 SYNOPSIS

 use Set::Similarity::Overlap;
 
 my $overlap = Set::Similarity::Overlap->new;
 my $similarity = $overlap->similarity('Photographer','Fotograf');
 
=head1 DESCRIPTION

=head2 Overlap coefficient

( A intersect B ) / min(A,B)

=head1 METHODS

L<Set::Similarity::Overlap> inherits all methods from L<Set::Similarity> and implements the
following new ones.

=head2 from_sets

  my $similarity = $object->from_sets(['a'],['b']);
 
This method expects two arrayrefs of strings as parameters. The parameters are not checked, thus can lead to funny results or uncatched divisions by zero.
 
If you want to use this method directly, you should take care that the elements are unique. Also you should catch the situation where one of the arrayrefs is empty (similarity is 0), or both are empty (similarity is 1).

=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Set-Similarity>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2014 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut




lib/Set/Similarity.pm
package Set::Similarity;

use strict;
use warnings;

our $VERSION = '0.026';

use Carp 'croak';

sub new {
  my $class = shift;

  # uncoverable condition false
  bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;
}

sub similarity {
  my ($self, $any1, $any2, $width) = @_;

  return $self->from_tokens(
    $self->_any($any1,$width),
    $self->_any($any2,$width)
  );
}

sub _any {
  my ($self, $any, $width) = @_;

  if (ref($any) eq 'ARRAY') {
    return $any;
  }
  elsif (ref($any) eq 'HASH') {
	return [grep { $any->{$_} } keys %$any];
  }
  elsif (ref($any)) {
    return [];
  }
  else {
    return [$self->ngrams($any,$width)];
  }
}

sub ngrams {
  my ($self, $word, $width) = @_;

  $width = 1 unless ($width && $width =~ m/^[1-9][0-9]*$/x);
  $word ||= '';

  return ($word) unless ($width <= length($word));

  return map {substr $word,$_,$width;} (0..length($word)-$width);
}

sub from_tokens {
  my ($self, $tokens1, $tokens2) = @_;

  return 1 if (!(scalar @$tokens1 || scalar @$tokens2));
  return 0 unless (scalar @$tokens1 && scalar @$tokens2 );

  return $self->from_sets(
    [$self->uniq($tokens1)],
    [$self->uniq($tokens2)],
  );
}

sub from_sets { croak 'Method "from_sets" not implemented in subclass' }

sub intersection {
  my %uniq;
  @uniq{@{$_[1]}} = ();
  scalar grep { exists $uniq{$_} } @{$_[2]};
}

sub uniq {
  my %uniq;
  @uniq{@{$_[1]}} = ();
  return keys %uniq;
}

sub combined_length {
  scalar(@{$_[1]}) + scalar(@{$_[2]});
}

sub min {
  (scalar(@{$_[1]}) < scalar(@{$_[2]}))
    ? scalar(@{$_[1]}) : scalar(@{$_[2]});
}

1;

__END__

=head1 NAME

Set::Similarity - similarity measures for sets

=begin html

<a href="https://travis-ci.org/wollmers/Set-Similarity"><img src="https://travis-ci.org/wollmers/Set-Similarity.png" alt="Set-Similarity"></a>
<a href='https://coveralls.io/r/wollmers/Set-Similarity?branch=master'><img src='https://coveralls.io/repos/wollmers/Set-Similarity/badge.png?branch=master' alt='Coverage Status' /></a>
<a href='http://cpants.cpanauthors.org/dist/Set-Similarity'><img src='http://cpants.cpanauthors.org/dist/Set-Similarity.png' alt='Kwalitee Score' /></a>
<a href="http://badge.fury.io/pl/Set-Similarity"><img src="https://badge.fury.io/pl/Set-Similarity.svg" alt="CPAN version" height="18"></a>

=end html

=head1 SYNOPSIS

 use Set::Similarity::Dice;

 # object method
 my $dice = Set::Similarity::Dice->new;
 my $similarity = $dice->similarity('Photographer','Fotograf');

 # class method
 my $dice = 'Set::Similarity::Dice';
 my $similarity = $dice->similarity('Photographer','Fotograf');

 # from 2-grams
 my $width = 2;
 my $similarity = $dice->similarity('Photographer','Fotograf',$width);

 # from arrayref of tokens
 my $similarity = $dice->similarity(['a','b'],['b']);

 # from hashref of features
 my $bird = {
   wings    => true,
   eyes     => true,
   feathers => true,
   hairs    => false,
   legs     => true,
   arms     => false,
 };
 my $mammal = {
   wings    => false,
   eyes     => true,
   feathers => false,
   hairs    => true,
   legs     => true,
   arms     => true,
 };
 my $similarity = $dice->similarity($bird,$mammal);

 # from arrayref sets
 my $bird = [qw(
   wings
   eyes
   feathers
   legs
 )];
 my $mammal = [qw(
   eyes
   hairs
   legs
   arms
 )];
 my $similarity = $dice->from_sets($bird,$mammal);

=head1 DESCRIPTION

This is the base class including mainly helper and convenience methods.

=head2 Overlap coefficient

( A intersect B ) / min(A,B)

=head2 Jaccard Index

The Jaccard coefficient measures similarity between sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets

( A intersect B ) / (A union B)

The Tanimoto coefficient is the ratio of the number of features common to both sets to the total number of features, i.e.

( A intersect B ) / ( A + B - ( A intersect B ) ) # the same as Jaccard

The range is 0 to 1 inclusive.

=head2 Dice coefficient

The Dice coefficient is the number of features in common to both sets relative to the average size of the total number of features present, i.e.

( A intersect B ) / 0.5 ( A + B ) # the same as sorensen

The weighting factor comes from the 0.5 in the denominator. The range is 0 to 1.

=head1 METHODS

All methods can be used as class or object methods.

=head2 new

  $object = Set::Similarity->new();

=head2 similarity

  my $similarity = $object->similarity($any1,$any1,$width);

C<$any> can be an arrayref, a hashref or a string. Strings are tokenized into n-grams of width C<$width>.

C<$width> must be integer, or defaults to 1.

=head2 from_tokens

  my $similarity = $object->from_tokens(['a','b'],['b']);

=head2 from_sets

  my $similarity = $object->from_sets(['a'],['b']);

Croaks if called directly. This method should be implemented in a child module.

=head2 intersection

  my $intersection_size = $object->intersection(['a'],['b']);

=head2 uniq

  my @uniq = $object->uniq(['a','b']);

Transforms an arrayref of strings into an array of unique elements.

=head2 combined_length

  my $set_size_sum = $object->combined_length(['a'],['b']);

=head2 min

  my $min_set_size = $object->min(['a'],['b']);

=head2 ngrams

  my @monograms = $object->ngrams('abc');
  my @bigrams = $object->ngrams('abc',2);

=head2 _any

  my $arrayref = $object->_any($any,$width);

=head1 SEE ALSO

L<Set::Similarity::Cosine>

L<Set::Similarity::Dice>

L<Set::Similarity::Jaccard>

L<Set::Similarity::Overlap>

L<Bag::Similarity> doing the same for bags or multisets.

L<Text::Levenshtein> for distance measures of strings, and a very overview of similar modules,

L<http://en.wikipedia.org/wiki/String_metric> for an overview of similarity measures.

L<Cluster::Similarity> for clusters.


=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Set-Similarity>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut@wollmersdorfer.atE<gt>

=begin html

<a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>

=end html

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2020 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

